### ISSUE TITLE
Returning NULL with an exception set from a py::init ctor

### ISSUE THREAD

**Issue by anntzer** [pybind/pybind11#1312]
## Issue description

It is apparently not allowable to return a null pointer with an exception set from a py::init constructor.  It would be nice if that was indeed possible (raising the exception on the Python side, of course).

## Reproducible example code

```

#include <pybind11/pybind11.h>

struct T { };

namespace py = pybind11;

PYBIND11_MODULE(python_example, m) {
    py::class_<T>(m, "T")
    .def(
      py::init(
        [](int x) -> T* {
          if (x == 0) {
            PyErr_SetString(PyExc_RuntimeError, "oops");
            return nullptr;
          } else {
            return new T{};
          }
        }));
}
```
calling `T(0)` yields "TypeError: pybind11::init(): factory function returned nullptr", when I was hoping for it to throw RuntimeError("oops").

Of course, in this case, I could just have thrown a c++ exception and rely on it being converted; the actual use case is when the constructor is calling back into the Python C-API, which may set an exception.

Originally posted on gitter.

**Comment by jagerman** [comment]
Changing `return nullptr` to `throw py::error_already_set()` should do what you want.

**Comment by jagerman** [comment]
Sounds good to me.  Perhaps a wording tweak:
"... initialized and just needs to be propagated to Python"
to:
"... initialized.  This exception is used to propagate a Python exception generated in C++ code back to Python."

**Comment by jagerman** [comment]
Closed via #1318.
